### ISSUE TITLE
"Define simple runfiles" cookbook example fails; runfiles in package-specific subdirectory

### ISSUE THREAD

**Issue by dfabulich** [bazelbuild/bazel#917]
https://github.com/dfabulich/runfiles-example follows the "Define simple runfiles" cookbook example in the `master` branch, but it's failing, because `data.txt` isn't in the current working directory; it's in a subdirectory, `pkg`.

```
$ bazel run pkg:e
INFO: Found 1 target...
Target //pkg:e up-to-date:
  bazel-bin/pkg/e
INFO: Elapsed time: 0.079s, Critical Path: 0.00s

INFO: Running command line: bazel-bin/pkg/e
cat: data.txt: No such file or directory
ERROR: Non-zero return code '1' from command: Process exited with status 1.
$ find bazel-bin/pkg/e.runfiles
bazel-bin/pkg/e.runfiles
bazel-bin/pkg/e.runfiles/MANIFEST
bazel-bin/pkg/e.runfiles/pkg
bazel-bin/pkg/e.runfiles/pkg/data.txt
bazel-bin/pkg/e.runfiles/pkg/e
```

**Comment by dfabulich** [comment]
The sample could be fixed by updating the `command` attribute to `cat pkg/data.txt` instead of just `cat data.txt`. But now consider the `multiple-packages` branch of the `runfiles-example` repository. In that example, `execute.bzl` hardcodes its command like this:

```
...
ctx.file_action(
    output=executable,
    content="cat data.txt",
    executable=True)
...
```

`foo/BUILD` and `bar/BUILD` look like this:

```
load('/pkg/execute', 'execute')

execute(
	name = 'foo',
	data = [':data.txt']
)
```

The `foo` and `bar` projects build successfully, but of course they don't work, because foo's data.txt is in `foo.runfiles/foo/data.txt` and bar's data.txt is in `bar.runfiles/bar/data.txt`.

In other words, the Skylark rule has no way of knowing in which subdirectory its runfiles will be deposited, so it has no way of referring to its own runfiles.

How do people workaround this?

**Comment by kchodorow** [comment]
In general you don't want any filenames in your rules.  In this example, you'd probably do something like:

```
content = "cat "
for d in ctx.attr.data:
  for f in d.files:
    content += "%s " % f.path
```

(untested). Check out the "Rule with attributes" example in the cookbook and http://bazel.io/docs/skylark/lib/File.html and http://bazel.io/docs/skylark/lib/Label.html.

**Comment by dfabulich** [comment]
I guess it makes sense to avoid explicit filenames in the rule itself, but what about directories?

For example, I have a script that looks for an unpredictable set of configuration files in `./config`. I thought I could use runfiles for this, but I can't, because the files won't be in `./config`, they'll be in `./foo/config` or `./bar/config`; my script has no way of knowing where to look for its configuration files. (Right?)

**Comment by ahumesky** [comment]
I think if you use ctx.files.data, it will give you all the files you want, e.g.:
```python
  file_paths = [f.path for f in ctx.files.data]
  ctx.file_action(
      output=executable,
      content="cat {data}".format(data = " ".join(file_paths)),
      executable=True)
```

**Comment by ahumesky** [comment]
You might try using glob() to collect all the files in the config directory, and passing that to the rule. Then all the files should be available to the executable. (Or maybe I misunderstand your question?)

**Comment by dfabulich** [comment]
But then how would I pass that data to the script? e.g. Maybe I'd write the list of files to a file, and store it in runfiles. But the whole problem is that the script doesn't know where its runfiles are, so it wouldn't be able to read that file, either.

**Comment by ahumesky** [comment]
Can you give some more detail about the layout of everything? Is it like this?

```
.
├── foo
│   ├── BUILD
│   ├── config
│   │   ├── a
│   │   ├── b
│   │   └── c
│   └── config.sh
├── execute.bzl
└── WORKSPACE
```

You should be able to do
```python
execute(
  name = "foo",
  data = glob(["config/*"])
)
```

and then pass the files like above

**Comment by damienmg** [comment]
No answer -> closing. Please ping back if you have more question
